### PR TITLE
Fixes #13986 - don't use image unless in image provisioning method

### PR DESCRIPTION
**Pull request by iNecas**

### CODE REVIEW

File: app/models/concerns/orchestration/compute.rb [hunk 1]
```diff
@@ -207,14 +207,20 @@ def find_image
   end
 
   def validate_compute_provisioning
-    return true unless image_build?
-    return true if ( compute_attributes.nil? or (compute_attributes[:image_id] || compute_attributes[:image_ref]).blank? )
-    img = find_image
-    if img
-      self.image = img
+    return true if compute_attributes.nil?
+    if image_build?
```
Comment (dLobatog) [inline app/models/concerns/orchestration/compute.rb:211]: Looks like this part of the if clause should be split up into another method

File: app/models/concerns/orchestration/compute.rb [hunk 2]
```diff
@@ -207,14 +207,20 @@ def find_image
...
+        self.image = img
+      else
+        failure(_("Selected image does not belong to %s") % compute_resource) and return false
+      end
     else
-      failure(_("Selected image does not belong to %s") % compute_resource) and return false
+      # don't send the image information to the compute resource unless using the image provisioning method
+      [:image_id, :image_ref].each { |image_key| compute_attributes.delete(image_key) }
     end
+    true
```
Comment (dLobatog) [inline app/models/concerns/orchestration/compute.rb]: No need to return anything? As long as it doesn't return false or `self` contains errors, the validation will pass

**Comment by dLobatog** [comment]
:+1: , thanks for the fix. Can you add some tests to ensure this bug doesn't creep back in?

File: app/models/concerns/orchestration/compute.rb (hunk 2 above)
Comment (iNecas) [inline app/models/concerns/orchestration/compute.rb]: Omitting the fact that the exit value of the orchestration methods has it's meeting bit us many times. I don't think that relying that neither of the conditional branches  will not have false as exit value (such as `logger.info` when level is set to `warn`: just an example) is a good idea.

File: app/models/concerns/orchestration/compute.rb (hunk 1 above)
Comment (iNecas) [inline app/models/concerns/orchestration/compute.rb:211]: Not sure I follow. Maybe if you suggest the name, I will understand better

File: app/models/concerns/orchestration/compute.rb (hunk 2 above)
Comment (iNecas) [inline app/models/concerns/orchestration/compute.rb]: anyway, I've remove the true here: not that painful as in the orchestration

**Comment by iNecas** [comment]
I've added tests for the new logic, as well as for checking the validation checks the images association

File: app/models/concerns/orchestration/compute.rb (hunk 1 above)
Comment (dLobatog) [inline app/models/concerns/orchestration/compute.rb:211]: It's alright, I just thought this part of the if could be an external method and called from here like:

``` ruby
if image_build?
  self.image = find_image 
else
  # delete image_id, image_ref from compute_attributes...
end
```

If it fails it's still fine. When you call `failure` etc... you're adding something to `.errors` so it'll fail. The method would also return false as `self.image` would be false.

Comment (dLobatog) [inline app/models/concerns/orchestration/compute.rb:211]: I'll just leave it as is, merge and I'll try to remember this PR whenever we modify this code again :smile:

**Comment by dLobatog** [comment]
Merged as 4c512942c3921cf8be696ed12cc588b1716b29d7, thanks @iNecas!

**Pull request closed without merging**